### ISSUE TITLE
How to handle temporary task errors?

### ISSUE THREAD

**Issue by AdamGleave** [ray-project/ray#2141]
How should Ray tasks behave in the face of temporary errors, specific to a worker? If a task raises an Exception, Ray will wrap this in a RayTaskError and never retry this task. This is usually what we want: but if the error is specific to the task running on that worker, this seems undesirable.

I came across a case like this today in the wild. When an AWS spot instance is scheduled for termination, all processes get killed, but -- in some cases -- not before a file share that my tasks depends on is unmounted. Accordingly, when an AWS spot instance is preempted, I get a FileNotFoundError for some percentage of my tasks whereas the rest are (correctly) retried.

Right now I can't see any way for my task to signal to Ray "this is a problem with this node, not the task itself". My workaround is to just kill the worker using sys.exit when this happens, before the exception can be caught by Ray! It then just handles it like a lost worker, which is OK. 

However, it'd be nice to have something like a NodeError exception that the task can raise so we get some reporting for what happened? Alternatively, Ray could automatically retry tasks up to some limit even after receiving an Exception (perhaps controllable as a parameter to ray.remote), although I'm weary of anything that will delay the reporting of errors.

**Comment by robertnishihara** [comment]
That's a really interesting question.

The workaround you're using makes sense. It doesn't seem like providing a NodeError exception would add value on top of your workaround (other than that it would feel cleaner and it wouldn't incur the delay from killing the worker).

Something like `@ray.remote(max_retries=5)` could make sense.

The implementation may be a little easier to hack together if the worker who was originally executing the task immediately does the retries, though it may be advantageous to have a different worker/machine execute the task the next time.

**Comment by stephanie-wang** [comment]
@AdamGleave, would an API that lets you retry the task that produced a particular object help? Not necessarily this API, but something along the lines of:

```
x = foo.remote()
while True:
  try:
    y = ray.get(x)
    break
  except:
    ray.recreate(x)  # Rerun the task that created x.

**Comment by AdamGleave** [comment]
I like the idea of letting the driver decide on task retries, this seems maximally flexible. In particular it would support applications where whether a task should be retried depends on the status of other tasks. For example, in parallel SGD it's fine to lose a few updates, but if you lost the majority of parameter updates in a batch you'd probably want to retry.

One problem: the exception raised by ray.get will always be a RayGetError. I imagine most users will want to retry tasks depending on the exception type (e.g. don't retry a TypeError which is something wrong with the code, but do retry an OSError). There doesn't seem to be an easy way to do this -- in fact, I think RayTaskError even throws away the original exception and just stores the traceback?

This approach also encourages all error handling taking place at the top level. For example, consider:
```
x = foo.remote()  # succeeds
y = bar.remote()  # fails on the first try
z = sum.remote(x, y)  # will always succeed for valid x & y
ray.get(z)  # fails
```
The error in ```y``` will only be apparent by ```ray.get(z)``` failing. The nice thing about this is we can do all the task-retrying logic in one place in the code. The nasty thing is we have to do it all in one place in the code, far away from where the error was actually generated. This will make it hard to do fine-grained error handling.        

In my particular use case this API feels a little awkward, since I'm storing Ray ObjectID's in nested dictionaries, and then calling ray.get in a map operation. A consequence of this is that I might block on the first ObjectID, while a second ObjectID (that I have not yet called ray.get on) has failed. In this case, I won't have an opportunity to retry the second task until the first has completed. I'm not sure how common this use case is though.

**Comment by mdr223** [comment]
@stephanie-wang I have a question related to fault tolerance which I detailed in this post on Discuss: https://discuss.ray.io/t/retry-task-w-different-resources/5772.

In brief, I'm wondering if it's possible to specify where (i.e. on what nodegroup) Ray should retry a task if its worker dies initially? This problem will come up frequently in problems where some inputs can cause a task/worker to run out of memory, and it would be nice to be able to direct the task to retry on a larger instance.

**Comment by danielgafni** [comment]
I'm having issue with using `ray.get()` on a list of refs containing nested ray tasks. 

I know the tasks can fail, and my code handles these exceptions internally. I am returning exceptions instead of raising them. 

This works as long as there is only one layer of ray tasks. As soon as there is a nested ray task which fails, `ray.get()` fails too. My top-level `try: except` catches the `RayTaskError` and tries to returns it, but it doesn't matter, because `ray.get` just fails anyway. 

This is problematic, because I don't have a way to connect the exception with a specific ref from the list of refs passed to `ray.get()`. I also don't want to retry this type of errors. 

Should there be an API for specifying how `ray.get` handles nested task failures?
